File: Napakalaki_Ruby/lib/specific_bad_consequence.rb
#encoding: utf-8

=begin
  Marco Manuel Fernández Pranno, 2015 (Granada)
  Para la asignatura PDOO (Universidad de Granada)
  Licencia GPLv3 a partir del 15/02/16, hasta entonces todos los derechos reservados.
=end 

class SpecificBadConsequence < BadConsequence
  def initialize(aText, someLevels, someSpecificVisibleTreasures, someSpecificHiddenTreasures)
    super(aText,someLevels)
    @specificVisibleTreasures = someSpecificVisibleTreasures
    @specificHiddenTreasures = someSpecificHiddenTreasures
  end
  
  def isEmpty()
    if((@specificHiddenTreasures.nil? || @specificHiddenTreasures.empty?) &&
           (@specificVisibleTreasures.nil? || @specificVisibleTreasures.empty? ) )
            return true
    end
      return false
  end
  
  def substractVisibleTreasure(t)
    unless( @specificVisibleTreasures.empty?)
      @specificVisibleTreasures.delete(t.tkind)
    end
  end
  
  def substractHiddenTreasure(t)
    unless( @specificHiddenTreasures.empty?)
      @specificHiddenTreasures.delete(t.tkind)
    end
  end
  
  def adjustToFitTreasureLists(visibles, hidden)
      
      if(@specificVisibleTreasures != nil)
        listaAjustadaVisibles = Array.new
      else
        listaAjustadaVisibles = nil
      end
      
      if(@specificHiddenTreasures != nil)
        listaAjustadaHidden = Array.new
      else
        listaAjustadaHidden = nil
      end
      
      visibleKind = visibles.collect{|t| t.tkind}
      hiddenKind = hidden.collect{|t| t.tkind}
      
      [TreasureKind::ARMOR,TreasureKind::SHOES,TreasureKind::HELMET,TreasureKind::ONEHAND,TreasureKind::BOTHHANDS].each do |tKind|
        if(@specificVisibleTreasures != nil)
          listaAjustadaVisibles = listaAjustadaVisibles + 
          [tKind]*[visibleKind.select{|t| t == tKind}.size, @specificVisibleTreasures.select{|t| t == tKind}.size].min
        end
        if(@specificHiddenTreasures != nil)
        listaAjustadaHidden = listaAjustadaHidden + 
          [tKind]*[hiddenKind.select{|t| t == tKind}.size, @specificHiddenTreasures.select{|t| t == tKind}.size].min
        end
      end 
      return SpecificBadConsequence.new(@text,0,listaAjustadaVisibles,listaAjustadaHidden)
  end
  
  attr_reader :specificVisibleTreasures, :specificHiddenTreasures
  public_class_method :new
  
  def to_s
    "--- Bad Consequence Especifico --- \n
    Texto: #{@text} \n Niveles: #{@levels} \n
    SpecificVisible: #{@specificVisibleTreasures} \n
    SpecificHidden: #{@specificHiddenTreasures} \n"
  end
end

File: Napakalaki_Java/src/NapakalakiGame/SpecificBadConsequence.java

package NapakalakiGame;

import static java.lang.Integer.min;
import java.util.ArrayList;

public class SpecificBadConsequence extends BadConsequence {
    
    protected ArrayList<TreasureKind> specificHiddenTreasures = null;
    protected ArrayList<TreasureKind> specificVisibleTreasures = null;

    public SpecificBadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden) {
        super(text, levels);
       
        this.specificVisibleTreasures = new ArrayList();
        this.specificHiddenTreasures = new ArrayList();
        this.specificVisibleTreasures = tVisible;
        this.specificHiddenTreasures = tHidden;
    }
    

    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> visibles, ArrayList<Treasure> hidden){
        BadConsequence bcAjustado;
        
        ArrayList<TreasureKind> listaAjustadaVisibles = null;
        ArrayList<TreasureKind> listaAjustadaHidden = null;

        for(TreasureKind tKind : TreasureKind.values()){
            int jugador = 0, badC = 0;
            
            if(specificVisibleTreasures != null){
                listaAjustadaVisibles = new ArrayList();
                for(TreasureKind t : specificVisibleTreasures){
                    if(t == tKind)
                        badC++;
                }

                for(Treasure t : visibles){
                    if(t.getType() == tKind)
                        jugador++;
                }

                for(int i = 0; i< min(jugador,badC); i++)
                    listaAjustadaVisibles.add(tKind);
            }
            
            jugador = 0;
            badC = 0;
            
            if(specificHiddenTreasures != null){
                listaAjustadaHidden = new ArrayList();
                for(TreasureKind t : specificHiddenTreasures){
                    if(t == tKind)
                        badC++;
                }

                for(Treasure t : hidden){
                    if(t.getType() == tKind)
                        jugador++;
                }

                for(int i = 0; i< min(jugador,badC); i++)
                    listaAjustadaHidden.add(tKind);
            }
        }

        bcAjustado = new SpecificBadConsequence(text, 0, listaAjustadaVisibles, listaAjustadaHidden);
        return bcAjustado;
    }
    
    @Override
    public boolean isEmpty(){
        if(((specificHiddenTreasures == null) && (specificVisibleTreasures == null)) || (specificHiddenTreasures.isEmpty() && specificVisibleTreasures.isEmpty()))
            return true;
        else
            return false;
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }

    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }

    @Override
    public void substractVisibleTreasure(Treasure t) {
        if (specificVisibleTreasures != null) {
            specificVisibleTreasures.remove(t.getType());
        }
    }

    @Override
    public void substractHiddenTreasure(Treasure t) {
        if (specificHiddenTreasures != null) {
            specificHiddenTreasures.remove(t.getType());
        }
    }

    @Override
    public String toString(){
        String description = "--- Mal Rollo Específico --- \n" + "Texto: " + text + "\n";
        
        if (levels != 0)
            description += "Pérdida de niveles: " + levels + "\n";
        
        if(specificVisibleTreasures != null)
            description += "Lista tesoros visibles: " + specificVisibleTreasures.toString() + "\n";
        if( specificHiddenTreasures != null)
            description += "Lista tesoros ocultos: " + specificHiddenTreasures.toString() + "\n";
                
        return description;
    }
}
